src/librustc_target/abi/call/wasm32.rs

```diff
@@ -1,16 +1,38 @@
-use crate::abi::call::{FnType, ArgType};
+use crate::abi::call::{FnType, ArgType, Uniform};
+use crate::abi::{HasDataLayout, LayoutOf, TyLayout, TyLayoutMethods};
 
-fn classify_ret_ty<Ty>(ret: &mut ArgType<'_, Ty>) {
-    ret.extend_integer_width_to(32);
+fn classify_ret_ty<'a, Ty, C>(cx: &C, ret: &mut ArgType<'a, Ty>)
+    where Ty: TyLayoutMethods<'a, C> + Copy,
+          C: LayoutOf<Ty = Ty, TyLayout = TyLayout<'a, Ty>> + HasDataLayout
+{
+    if ret.layout.is_aggregate() {
+        if let Some(unit) = ret.layout.homogeneous_aggregate(cx).unit() {
+            let size = ret.layout.size;
+            if unit.size == size {
```

Comment:
Looking at https://github.com/llvm/llvm-project/blob/d01b4a786271c9ce9df43add56ec2e3fb40ce920/clang/lib/CodeGen/TargetInfo.cpp#L791-L832, this does seem correct, but Clang appears to do this for *both* arguments and return, not just return.